### ISSUE TITLE
Breaking Change in 5.0.29 with `artisan migrate:reset`

### ISSUE THREAD

**Issue by pvormittag** [laravel/framework#8670]
### TL;DR

This [commit](https://github.com/laravel/framework/commit/5792da9476c07c8e89dbb8d4e7fd4ab7b0bf87f7) appears to introduce a breaking change where migrations are reversed in an order which does not respect the batch number. I noticed the issue while running `php artisan migrate:reset`.

### The Issue

I have a migrations table as follows:

| migration                                                    | batch |
|--------------------------------------------------------------|-------|
| 2015_01_01_000001_create_sapling_environment_variables_table | 1     |
| 2015_01_01_000001_create_oauth_sessions_table                | 2     |
| 2015_01_01_000001_create_initial_transit_tables              | 3     |
| 2015_01_01_000001_create_initial_core_resident_tables        | 4     |
| 2015_03_01_000001_create_initial_core_portfolio_tables       | 5     |
| 2015_01_01_000001_create_initial_core_keychain_tables        | 6     |
| 2015_01_01_000001_create_initial_core_assets_tables          | 7     |
| 2015_01_01_000001_create_initial_core_content_blocks_tables  | 8     |
| 2015_01_01_000001_create_initial_transit_foreign_keys        | 9     |
| 2015_01_01_000001_create_initial_core_resident_foreign_keys  | 10    |

When I run `php artisan migrate:reset --pretend` in Laravel `5.0.28`, I recieve the following output, which maps correctly to the migrations table by which each migration should run and their batch number.

```bash
CreateInitialCoreResidentForeignKeys: alter table `core_resident_portfolios` drop foreign key core_resident_portfolios_core_portfolio_id_foreign
CreateInitialCoreResidentForeignKeys: alter table `core_resident_events` drop foreign key core_resident_events_portfolio_id_foreign
CreateInitialCoreResidentForeignKeys: alter table `core_resident_offers` drop foreign key core_resident_offers_portfolio_id_foreign
CreateInitialCoreResidentForeignKeys: alter table `core_resident_packages` drop foreign key core_resident_packages_portfolio_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_social_mashup_handle` drop foreign key transit_screens_social_m_screen_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_social_mashup_handle` drop foreign key transit_screens_social_m_provider_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_social_mashup` drop foreign key transit_screens_social_m_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_social_mashup` drop foreign key transit_screens_social_m_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_social_mashup` drop foreign key transit_screens_social_m_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_weather` drop foreign key transit_screens_weather_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_weather` drop foreign key transit_screens_weather_service_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_weather` drop foreign key transit_screens_weather_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_weather` drop foreign key transit_screens_weather_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_video` drop foreign key transit_screens_video_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_video` drop foreign key transit_screens_video_service_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_video` drop foreign key transit_screens_video_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_video` drop foreign key transit_screens_video_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_tweet` drop foreign key transit_screens_tweet_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_tweet` drop foreign key transit_screens_tweet_twitter_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_tweet` drop foreign key transit_screens_tweet_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_tweet` drop foreign key transit_screens_tweet_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_transit` drop foreign key transit_screens_transit_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_transit` drop foreign key transit_screens_transit_service_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_transit` drop foreign key transit_screens_transit_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_transit` drop foreign key transit_screens_transit_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_package_notification` drop foreign key transit_screens_package__layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_package_notification` drop foreign key transit_screens_package__service_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_package_notification` drop foreign key transit_screens_package__image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_package_notification` drop foreign key transit_screens_package__image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_offer` drop foreign key transit_screens_offer_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_offer` drop foreign key transit_screens_offer_service_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_offer` drop foreign key transit_screens_offer_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_offer` drop foreign key transit_screens_offer_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_message` drop foreign key transit_screens_message_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_message` drop foreign key transit_screens_message_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_message` drop foreign key transit_screens_message_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_message` drop foreign key transit_screens_message_image_foreground_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_event` drop foreign key transit_screens_event_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_event` drop foreign key transit_screens_event_service_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_event` drop foreign key transit_screens_event_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_event` drop foreign key transit_screens_event_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_providers` drop foreign key transit_providers_portfolio_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_providers` drop foreign key transit_providers_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_slides` drop foreign key transit_slides_channel_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_channels` drop foreign key transit_channels_portfolio_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_portfolios` drop foreign key transit_portfolios_core_portfolio_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_portfolios` drop foreign key transit_portfolios_theme_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_portfolios` drop foreign key transit_portfolios_alert_channel_id_foreign
CreateInitialCoreContentBlocksTables: drop table `core_content_blocks`
CreateInitialCoreAssetsTables: drop table `core_assets`
CreateInitialCoreKeychainTables: drop table `core_keychain_keys`
CreateInitialCorePortfolioTables: drop table `core_portfolios`
CreateInitialCoreResidentTables: drop table `core_resident_packages`
CreateInitialCoreResidentTables: drop table `core_resident_offers`
CreateInitialCoreResidentTables: drop table `core_resident_events`
CreateInitialCoreResidentTables: drop table `core_resident_portfolios`
CreateInitialTransitTables: drop table `transit_screens_social_mashup_handle`
CreateInitialTransitTables: drop table `transit_screens_social_mashup`
CreateInitialTransitTables: drop table `transit_screens_weather`
CreateInitialTransitTables: drop table `transit_screens_video`
CreateInitialTransitTables: drop table `transit_screens_tweet`
CreateInitialTransitTables: drop table `transit_screens_transit`
CreateInitialTransitTables: drop table `transit_screens_package_notification`
CreateInitialTransitTables: drop table `transit_screens_offer`
CreateInitialTransitTables: drop table `transit_screens_message`
CreateInitialTransitTables: drop table `transit_screens_event`
CreateInitialTransitTables: drop table `transit_screen_layouts`
CreateInitialTransitTables: drop table `transit_providers`
CreateInitialTransitTables: drop table `transit_slides`
CreateInitialTransitTables: drop table `transit_channels`
CreateInitialTransitTables: drop table `transit_themes`
CreateInitialTransitTables: drop table `transit_portfolios`
CreateOauthSessionsTable: drop table `oauth_sessions`
CreateSaplingEnvironmentVariablesTable: drop table `sapling_environment_variables`
```

However, when I run `php artisan migrate:reset --pretend` in Laravel `5.0.29`, I recieve a different output, which appears random and doesn't map back to the migrations table and the batch number. I suspect this [commit](https://github.com/laravel/framework/commit/5792da9476c07c8e89dbb8d4e7fd4ab7b0bf87f7) is the reason for this. In my case, ordering by the name of the migration without first ordering by the batch column will not produce the correct order, since the timestamp on each migration is identical. The reason they are identical is likely irrelevant, however its due to each migration being in a different composer package. A possible solution would be to order first on batch number then on migration name.

```bash
CreateInitialCorePortfolioTables: drop table `core_portfolios`
CreateSaplingEnvironmentVariablesTable: drop table `sapling_environment_variables`
CreateOauthSessionsTable: drop table `oauth_sessions`
CreateInitialTransitTables: drop table `transit_screens_social_mashup_handle`
CreateInitialTransitTables: drop table `transit_screens_social_mashup`
CreateInitialTransitTables: drop table `transit_screens_weather`
CreateInitialTransitTables: drop table `transit_screens_video`
CreateInitialTransitTables: drop table `transit_screens_tweet`
CreateInitialTransitTables: drop table `transit_screens_transit`
CreateInitialTransitTables: drop table `transit_screens_package_notification`
CreateInitialTransitTables: drop table `transit_screens_offer`
CreateInitialTransitTables: drop table `transit_screens_message`
CreateInitialTransitTables: drop table `transit_screens_event`
CreateInitialTransitTables: drop table `transit_screen_layouts`
CreateInitialTransitTables: drop table `transit_providers`
CreateInitialTransitTables: drop table `transit_slides`
CreateInitialTransitTables: drop table `transit_channels`
CreateInitialTransitTables: drop table `transit_themes`
CreateInitialTransitTables: drop table `transit_portfolios`
CreateInitialTransitForeignKeys: alter table `transit_screens_social_mashup_handle` drop foreign key transit_screens_social_m_screen_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_social_mashup_handle` drop foreign key transit_screens_social_m_provider_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_social_mashup` drop foreign key transit_screens_social_m_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_social_mashup` drop foreign key transit_screens_social_m_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_social_mashup` drop foreign key transit_screens_social_m_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_weather` drop foreign key transit_screens_weather_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_weather` drop foreign key transit_screens_weather_service_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_weather` drop foreign key transit_screens_weather_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_weather` drop foreign key transit_screens_weather_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_video` drop foreign key transit_screens_video_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_video` drop foreign key transit_screens_video_service_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_video` drop foreign key transit_screens_video_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_video` drop foreign key transit_screens_video_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_tweet` drop foreign key transit_screens_tweet_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_tweet` drop foreign key transit_screens_tweet_twitter_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_tweet` drop foreign key transit_screens_tweet_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_tweet` drop foreign key transit_screens_tweet_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_transit` drop foreign key transit_screens_transit_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_transit` drop foreign key transit_screens_transit_service_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_transit` drop foreign key transit_screens_transit_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_transit` drop foreign key transit_screens_transit_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_package_notification` drop foreign key transit_screens_package__layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_package_notification` drop foreign key transit_screens_package__service_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_package_notification` drop foreign key transit_screens_package__image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_package_notification` drop foreign key transit_screens_package__image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_offer` drop foreign key transit_screens_offer_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_offer` drop foreign key transit_screens_offer_service_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_offer` drop foreign key transit_screens_offer_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_offer` drop foreign key transit_screens_offer_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_message` drop foreign key transit_screens_message_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_message` drop foreign key transit_screens_message_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_message` drop foreign key transit_screens_message_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_message` drop foreign key transit_screens_message_image_foreground_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_event` drop foreign key transit_screens_event_layout_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_event` drop foreign key transit_screens_event_service_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_event` drop foreign key transit_screens_event_image_preview_thumbnail__foreign
CreateInitialTransitForeignKeys: alter table `transit_screens_event` drop foreign key transit_screens_event_image_background_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_providers` drop foreign key transit_providers_portfolio_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_providers` drop foreign key transit_providers_key_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_slides` drop foreign key transit_slides_channel_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_channels` drop foreign key transit_channels_portfolio_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_portfolios` drop foreign key transit_portfolios_core_portfolio_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_portfolios` drop foreign key transit_portfolios_theme_id_foreign
CreateInitialTransitForeignKeys: alter table `transit_portfolios` drop foreign key transit_portfolios_alert_channel_id_foreign
CreateInitialCoreResidentTables: drop table `core_resident_packages`
CreateInitialCoreResidentTables: drop table `core_resident_offers`
CreateInitialCoreResidentTables: drop table `core_resident_events`
CreateInitialCoreResidentTables: drop table `core_resident_portfolios`
CreateInitialCoreResidentForeignKeys: alter table `core_resident_portfolios` drop foreign key core_resident_portfolios_core_portfolio_id_foreign
CreateInitialCoreResidentForeignKeys: alter table `core_resident_events` drop foreign key core_resident_events_portfolio_id_foreign
CreateInitialCoreResidentForeignKeys: alter table `core_resident_offers` drop foreign key core_resident_offers_portfolio_id_foreign
CreateInitialCoreResidentForeignKeys: alter table `core_resident_packages` drop foreign key core_resident_packages_portfolio_id_foreign
CreateInitialCoreKeychainTables: drop table `core_keychain_keys`
CreateInitialCoreContentBlocksTables: drop table `core_content_blocks`
CreateInitialCoreAssetsTables: drop table `core_assets`
```

**Comment by GrahamCampbell** [comment]
Thank you for reporting. We will revert this change very shortly. :)

**Comment by GrahamCampbell** [comment]
5.0.30 has been tagged.

**Comment by pvormittag** [comment]
Great!

**Comment by willrowe** [comment]
@pvormittag while `artisan migrate:reset` should respect the batch numbers, the names of your migrations weren't generated correctly. Did you enter them manually? If you were to run `artisan migrate:reset` and subsequently `artisan migrate`, it would not migrate them in the original order, but rather in alphabetical order of the file names, since it's expecting that you generated them chronologically.

**Comment by pvormittag** [comment]
@willrowe The name of the migration shouldn't matter since each migration is associated with a different batch number.

As I mentioned in the issue, the migrations are spread across many composer packages, which may or may not depend on one another. Within each package, there may be many migration files, and those do follow a naming convention of 2015_01_01_000001_*, 2015_01_01_000002_*, etc, it just happens not to be the case in my issue above... I realize it's idiomatic for Laravel to generate migration names using the current time, however, as long as an ordered sequence is maintained in the format Laraval expects, weather or not the name contains a "true" timestamp is irrelevant.

**Comment by willrowe** [comment]
Thanks for the explanation. I agree that a possible solution to both problems would be to just sort by batch number first.

**Comment by reinink** [comment]
For what it's worth, I think this issue still exists. I don't have time to dig into what's going wrong here, but I keep getting issues with `migrate:reset` where migrations with a LOWER batch number run first, which obviously causes big issues. I'd love to help dig into what is causing this, but for now I just wanted to post on here that I'm still having this problem, just in case others are as well.
